sdk/core/azure-core/src/main/java/com/azure/core/implementation/FileWriteSubscriber.java

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import reactor.core.publisher.MonoSink;
+import reactor.core.publisher.Operators;
+
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousFileChannel;
+import java.nio.channels.CompletionHandler;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Subscriber that writes a stream of {@link ByteBuffer ByteBuffers} to a file.
+ */
+@SuppressWarnings("ReactiveStreamsSubscriberImplementation")
+public final class FileWriteSubscriber implements Subscriber<ByteBuffer> {
+
+    // volatile ensures that writes to these fields by one thread will be immediately visible to other threads.
+    // An I/O pool thread will write to isWriting and read isCompleted,
+    // while another thread may read isWriting and write to isCompleted.
+    private volatile boolean isWriting = false;
+    private volatile boolean isCompleted = false;
+
+    private final AsynchronousFileChannel fileChannel;
+    private final AtomicLong position;
+    private final MonoSink<Void> emitter;
+
+    private Subscription subscription;
+
+    public FileWriteSubscriber(AsynchronousFileChannel fileChannel, long position, MonoSink<Void> emitter) {
+        this.fileChannel = fileChannel;
+        this.position = new AtomicLong(position);
+        this.emitter = emitter;
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        // Only set the Subscription if one has not been previously set.
+        // Any additional Subscriptions will be cancelled.
+        if (Operators.validate(this.subscription, s)) {
+            subscription = s;
+
+            s.request(1);
+        }
+    }
+
+    @Override
+    public void onNext(ByteBuffer bytes) {
+        try {
+            if (isWriting) {
+                onError(new IllegalStateException("Received onNext while processing another write operation."));
+            } else {
+                write(bytes);
+            }
+        } catch (Throwable throwable) {
+            // If writing has an error, and it isn't caught, there is a possibility for it to deadlock the reactive
+            // stream. Catch the exception and propagate it manually so that doesn't happen.
+            onError(throwable);
+        }
+    }
+
+    private void write(ByteBuffer bytes) {
+        isWriting = true;
+
+        fileChannel.write(bytes, position.get(), bytes, new CompletionHandler<Integer, ByteBuffer>() {
+            @Override
+            public void completed(Integer result, ByteBuffer attachment) {
+                position.addAndGet(result);
+
+                if (bytes.hasRemaining()) {
```

Comment:
I decided to use the object pointer given in the method call instead of what we being passed back by the `completed` method as it is dependent on the underlying file implementation to pass this back properly. I was running into some testing issue when I was previously using `attachment`.

I could update this to use a two-phase approach, if `attachment != null` use `attachment`, otherwise use `bytes`. Thoughts?

Reply:
Let's continue to use `bytes` since we are noticing issues with using `attachment`